src/theory/tools/wp/bir_htScript.sml

```diff
@@ -0,0 +1,614 @@
+  ``!prog bl st.
+      ?l' c' st'.
+        bir_exec_block prog bl st = (l', c', st')``,
+
+Cases_on `prog` >>
+FULL_SIMP_TAC std_ss [bir_exec_block_def] >>
+REPEAT STRIP_TAC >>
+ASSUME_TAC (SPECL [``bl.bb_statements:'a bir_stmt_basic_t list``,
+                   ``[]: 'a list``, ``0:num``, ``st:bir_state_t``]
+                  bir_exec_stmtsB_exists) >>
+FULL_SIMP_TAC std_ss [LET_DEF] >>
+Cases_on `bir_state_is_terminated st'` >> (
+  FULL_SIMP_TAC std_ss []
+)
+);
+
+val bir_prog_not_assume_never_assviol =
+  store_thm("bir_prog_not_assume_never_assviol",
+  ``!prog st n bl ol nstep npc st'.
+    (bir_get_current_block prog st.bst_pc = SOME bl) ==>
+    bir_prog_has_no_assumes prog ==>
+    (st.bst_status <> BST_AssumptionViolated) ==>
+    (bir_exec_block_n prog st n =
+       (ol, nstep, npc, st')
+    ) ==>
+    (st'.bst_status <> BST_AssumptionViolated)``,
+
+Induct_on `n` >| [
+  FULL_SIMP_TAC std_ss [bir_exec_block_n_def,
+                        bir_exec_steps_GEN_REWR_no_steps,
+                        valOf_BER_Ended_def],
+
+  REPEAT STRIP_TAC >>
+  ASSUME_TAC (SPECL [``prog:'a bir_program_t``,
+		     ``bl:'a bir_block_t``,
+		     ``st:bir_state_t``] bir_exec_block_exists) >>
+  FULL_SIMP_TAC std_ss [] >>
+  (* How to best swap variable names? *)
+  rename1 `st'''.bst_status = BST_AssumptionViolated` >>
+  rename1 `bir_exec_block prog bl st = (l',c',st')` >>
+  rename1 `st''.bst_status = BST_AssumptionViolated` >>
+  PAT_X_ASSUM ``!prog st bl ol nstep npc st'. blah``
+          (fn thm => ASSUME_TAC (SPECL [``prog: 'a bir_program_t``,
+                                        ``st':bir_state_t``] thm)
+          ) >>
+  IMP_RES_TAC test_lemma >>
+  IMP_RES_TAC bir_block_not_assume_never_assviol >>
+  IMP_RES_TAC bir_exec_block_n_block >>
+  PAT_X_ASSUM ``!n. blah``
+          (fn thm => ASSUME_TAC
+            (SPECL [``n:num``] thm)
+          ) >>
+  REV_FULL_SIMP_TAC std_ss [LET_DEF] >>
+  Cases_on `0 < c' /\
+            bir_state_COUNT_PC (F,(\pc. pc.bpc_index = 0)) st'`
+    >> (
+      FULL_SIMP_TAC std_ss [] >>
+      FULL_SIMP_TAC std_ss []
+   ) >> (
+    Cases_on `~bir_state_is_terminated st'` >| [
+      IMP_RES_TAC bir_exec_block_new_block_pc >>
+      FULL_SIMP_TAC std_ss [optionTheory.IS_SOME_EXISTS] >>
+      FULL_SIMP_TAC (std_ss++holBACore_ss) [] >>
+      PAT_X_ASSUM ``(ol,nstep,npc,st'') = blah``
+	      (fn thm => ASSUME_TAC
+		(GSYM thm)
+	      ) >>
+      (* TODO: The below finish is ugly... *)
+      Cases_on `(bir_exec_block_n prog st' n)` >>
+      Cases_on `r` >>
+      Cases_on `r'` >>
+      FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+      FULL_SIMP_TAC (std_ss++holBACore_ss) [] >>
+      IMP_RES_TAC bir_exec_block_n_REWR_TERMINATED >>
+      PAT_X_ASSUM ``!p n. blah``
+	      (fn thm => ASSUME_TAC
+		(SPECL [``prog:'a bir_program_t``, ``n:num``] thm)
+	      ) >>
+      FULL_SIMP_TAC (std_ss++holBACore_ss) []
+    ]
+  ) 
+]
+);
+(* Formulated in terms of bir_exec_to_labels: *)
```

Comment:
If you need a comment to explain the name, maybe the naming is bad? :) Maybe you can include "exec" and "exec_to_labels" in the names?

Reply:
Okay, I am now renaming this and the previous theorem to reference the execution functions more explicitly.